### PR TITLE
Correctly handle a retransmitted ClientHello

### PR DESCRIPTION
**Pull request by mattcaswell**
If we receive a ClientHello and send back a HelloVerifyRequest, we need
to be able to handle the scenario where the HelloVerifyRequest gets lost
and we receive another ClientHello with the message sequence number set to
0.

Fixes https://github.com/openssl/openssl/issues/18635

### CODE REVIEW

**Comment by mattcaswell** [comment]
Note there is a trivial conflict when cherry-picking back to 1.1.1 which I will fix during merge.

**Review by t8m — APPROVED** [review]

**Comment by mattcaswell** [comment]
I rebased this to resolve a conflict with master.

Still looking for a second reviewer. Ping @openssl/committers.

**Review by t8m — APPROVED** [review]

File: ssl/statem/statem_dtls.c [hunk 1]
```diff
@@ -496,31 +496,72 @@ static int dtls1_retrieve_buffered_fragment(SSL_CONNECTION *s, size_t *len)
...
+                dtls1_hm_fragment_free(frag);
+                pitem_free(item);
+                item = NULL;
+                frag = NULL;
+            } else {
+                /*
+                 * We have fragments for a ClientHello without a cookie,
+                 * even though we have sent a HelloVerifyRequest. It is possible
+                 * that the HelloVerifyRequest got lost and this is a
+                 * retranmission of the original ClientHello
```
Comment (hlandau) [inline ssl/statem/statem_dtls.c:535]: nit: typo, retransmission

**Review by hlandau — CHANGES_REQUESTED** [review]

**Comment by mattcaswell** [comment]
@hlandau Typo fixed. Please take another look.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Meanwhile the 1.1.1 got to the state when only security fixes are supposed to be done there.

**Review by hlandau — APPROVED** [review]

**Comment by hlandau** [comment]
Does this go to 1.1.1?

**Comment by mattcaswell** [comment]
> Does this go to 1.1.1?

No - its not a security fix. Removing the 1.1.1 branch label.

**Comment by hlandau** [comment]
It's now been more than 24 hours. Merging.

**Comment by hlandau** [comment]
Merged to master and 3.0. Thank you.

A merge conflict while cherry picking to 3.0 was deemed trivial enough to fix during merging. (Adjacent line to a changed line had changed so the diff algorithm wasn't happy, but no actual diff changes.)

**Pull request closed without merging**